Pin ESP32 to core 1 once again ftw

breedx2 opened the pull request:
Hi friends. New contributor, but pretty experienced...and hopefully I have done my homework. Phew, there is a lot of context/history in a very small PR here... :)

Originally we [had some latency issues](https://github.com/micropython/micropython/issues/3894) that were resolved by [pinning micropython to a core](https://github.com/micropython/micropython/pull/4943), specifically opposite of the other stuff going on (BLE/Wifi).  That "other stuff" could introduce latency or jitter in timer or interrupt callback delivery.  All good, cool workaround.

Some of this seemed to be rooted in an upstream bug in IDF < 4.2, so we [pinned everything to core0](https://github.com/micropython/micropython/issues/5489) in the esp32 port.  [Related.](https://github.com/micropython/micropython/pull/6343).  There were comments to indicate that when this upstream situation was resolved we could undo this...

Time passed.  A major branch got integrated.  The core pinning stayed or otherwise got left behind as IDF improved (I couldn't track down the root cause in IDF).

Fast forward to the future...and I boot up a project based on the 1.15 release, and I'm doing what I _think_ to be a pretty simple thing -- 1ms pin toggles based on a `Timer`.  It's rock solid, until I turn on wifi...at which point it becomes unstable (even with no traffic).  I make a pretty basic [test case that I can put on the oscilloscope](https://github.com/breedx2/led_sign_2020/blob/the_unbinding/software/src/test/test2.py) and it is solid without wifi, significantly jittery with wifi on.

LOTS of digging ensues, and I end up finding the above-related issues.  

I checked out the main branch and built with IDF 4.2.1, same problem.  I made the change in this PR and _presto_ the problem goes away and the latency is solid again.  Woohoo!  Let's do this!

Sadly, my target device doesn't leverage BLE so I'm able to verify that this doesn't regress the original issue...but I _can_ verify that `Timer()` instances now deliver with negligible jitter when WiFi is on.  Can others with BLE-based targets confirm that this patch doesn't impact them?  Are we ready to roll with pinning to core 1 again please, cuz this 100% helps me.  

Thanks thanks.

dpgeorge commented on ports/esp32/mphalport.h:
```diff
@@ -36,10 +36,15 @@
+// and avoid the Wifi/BLE timing problems on the same core.
+// Best effort here to remain backwards compatible in rare version edge cases...
+// See https://github.com/micropython/micropython/issues/5489 for history
+#if ESP_IDF_VERSION >= ESP_IDF_VERSION_VAL(4, 2, 1)
```
Will it work with IDF 4.2?

breedx2 replied:
Hey @dpgeorge good call out...I just tried with 4.2.1 because it was the latest...but I wlll try 4.2 tonight and report back.  If it works in my use case I will update this PR.

breedx2 replied:
@dpgeorge I can confirm that my problem is also resolved with this patch and IDF 4.2.0, so I have update the PR to match.  Thanks!

breedx2 committed:
esp32: Pin MicroPython to core 1 again.

This follows up on #5489, where we changed the esp32 core pinning
to core 0 in order to work around an issue with IDF < 4.2.0.
Now that IDF > 4.2.0 is available, we allow pinning back to core
1, which eliminates some problematic callback latency with
wifi enabled.

dpgeorge commented:
I tested BLE with this PR and it does not work, events get lost.  It starts working again if NimBLE is pinned to core 1.

dpgeorge commented:
It's probably OK to move NimBLE to core 1 as well.  Otherwise wait for #7046 which may resolve the issue of pinning NImBLE.

breedx2 committed:
esp32/boards: Pin NimBLE to core 1 if IDF > 4.2.

This ensures that NimBLE is on the same core as Micropython, even
in IDF versions above 4.2.

breedx2 commented:
@dpgeorge Oh yes, I initially missed that NimBLE pinning is in another file.  Keeping them on the same core for now seems like a good agreeable choice, so thanks for the advice.  I have made the corresponding change, so please let me know if there is anything else to address in this PR.

dpgeorge commented on ports/esp32/boards/sdkconfig.ble:
```diff
@@ -9,9 +9,15 @@ CONFIG_BT_NIMBLE_ENABLED=y
-# See https://github.com/micropython/micropython/issues/5489
+# Prior to IDF 4.2+, we needed to pin NimBLE to core 0 for 
+# synchronization with the ringbuffer. This 
+#if ESP_IDF_VERSION >= ESP_IDF_VERSION_VAL(4, 2, 0)
```
I'm pretty sure #if's are not supported in Kconfig files... I guess the only way to deal with this is to do the selection in the CMakeLists.txt file, which is getting tricky.

breedx2 replied:
I wondered.  I'll just pin to 1 then. :bowing_man:

dpgeorge replied:
But now it won't work with IDF < 4.2.

breedx2 replied:
What's the path forward then?  Does my original intent to pin opposite of WiFi make things _worse_ for BLE and we just need to wait for that other thing....or what do you think we should do now?

dpgeorge replied:
Two options:
1. wait for #7046 (it's very close) and then keep NimBLE on core 0
2. make CMakeLists.txt include the right sdkconfig fragment to pin NimBLE depending on the IDF version

Option (2) is not that difficult, but might anyway become obsolete once option (1) is in.

breedx2 replied:
Ok, with that in mind, I can put NimBLE back on core0 and we can wait this one out.  :)  

Thanks again for the reviews and help and advice here.

breedx2 committed:
esp32/boards: Fix/undo misunderstood macros

Whoops, probably should not have expected c macros to work
in kconfig files.  This undoes that and just plain pins
to core 1.

breedx2 committed:
revert to leave BLE on core 0 (and wait for #7046)

dpgeorge commented:
It looks like #7046 might take a while, so I've merged this PR, with an addition which makes sure NimBLE is on the same core as MicroPython.   See 357078504d0d2c3a3f06ec4bd69e0ab3bcf9e016

See also follow up commit 648656dbbd841aae129628487a71037236aac739

Hopefully this improves timing based things on esp32!
